### PR TITLE
Extend models submodule

### PR DESCRIPTION
**Pull request by fbanning**
Related to #258 

Did I miss anything?
Happy to receive feedback if anything should be changed. (But please be kind, it's my first PR for an open source project.)

### CODE REVIEW

**Comment by Libbum** [comment]
A good start! Just a few more things to take care of.

- [ ] Alter the files under `examples/` so that our documentation updates itself. Just a simple one liner will do it. This is the change done to [forrest_fire](https://github.com/JuliaDynamics/Agents.jl/commit/bb9780bc9a9926da0f1329f59de98e2a6db8f76d#diff-603bac5915bb7504a05320eb1f87e647) for example.
- [ ] The tests are failing because you're re-defining `agent_step!` and a few other things in multiple files. Here's the [forrest_fire](https://github.com/JuliaDynamics/Agents.jl/commit/bb9780bc9a9926da0f1329f59de98e2a6db8f76d#diff-bc279cd292aec31c06934d4c7c5928f1) model diff. See how we name the function `forest_step!`? We also have `schelling_step!` etc.

Once those are fixed, I'll need to run the documentation build locally to check & verify the changes are all correct. For the most part it looks good though, thanks!

**Comment by fbanning** [comment]
Great, I'll fix these things on monday. 👍

**Comment by fbanning** [comment]
Considering the renaming of the stepping functions:
I would propose to rename all of them following the pattern `[name]_agent_step!` and `[name]_model_step!`. This will help to make clear which function is which because otherwise you would just have e.g. `forest_step!` (unclear if agent or model step).
Are you OK with this?

**Comment by fbanning** [comment]
One question about https://github.com/fbanning/Agents.jl/blob/Extend-Models-submodule/src/models/wright-fisher.jl : 
Is it really necessary to have those two functions with different names `*_neutral` and `*_selection` ? I think it might be possible to just define the function twice with the same name - once with and once without `:trait`as a parameter.
If that's correct, this could simplify the module a bit.

**Comment by Datseris** [comment]
@kavir1698 knows about the evolution model

**Comment by Libbum** [comment]
> I would propose to rename all of them following the pattern `[name]_agent_step!` and `[name]_model_step!`

This is a good idea. Being explicit is useful here, users can re-name these functions anyway they like anyway. So go ahead with this.

**Comment by fbanning** [comment]
Great. I've already integrated that in the commit 069389629574bdf84b6787083f8a825a4b23c138.

How come the Travis checks are still failing?

**Comment by Libbum** [comment]
I don't have access to push to your branch. But the (current) Travis failure is due to the inclusion of `Distributions` in the SIR model. I think it's fine to add this as a dependency to the project.

Jump into the julia repl in your branch, 
```julia
] activate .
add Distributions
```
Then add the altered `Project.toml` and push it up. That should fix the current problem.

**Comment by fbanning** [comment]
Ah OK, didn't think of that. I haven't worked with Travis before and didn't know where to look.

File: src/models/daisyworld.jl [hunk 1]
```diff
@@ -0,0 +1,163 @@
+using Random
+import StatsBase
+import DrWatson: @dict
```
Comment (Libbum) [inline src/models/daisyworld.jl:3]: This is the current Travis error. It's just a convenience function, so let's drop this dependency.

File: src/models/daisyworld.jl [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        solar_change = 0.005,
+        solar_luminosity = 1.0, # initial luminosity
+        scenario = :default,
+        seed = 165
+    )
+
+    Random.seed!(seed)
+    space = GridSpace(griddims, moore = true, periodic = true)
+    properties = @dict max_age surface_albedo solar_luminosity solar_change scenario
+    properties[:tick] = 0
```
Comment (Libbum) [inline src/models/daisyworld.jl:54]:
```julia
properties = Dict(max_age => max_age,
 surface_albedo => surface_albedo,
 solar_luminosity => solar_luminosity,
 solar_change => solar_change,
 scenario => scenario,
 tick => 0)
```

**Comment by Datseris** [comment]
The **documentation** dependencies are completely independent from the actual Agents.jl dependencies (see the file in `docs/Project.toml`) and there we can add any dependency, and I don't think we should remove any, especially the very handy `@dict` macro.

File: src/models/sir.jl [hunk 3]
```diff
@@ -0,0 +1,182 @@
+using Distributions: Poisson, DiscreteNonParametric
+using DrWatson: @dict
```
Comment (Libbum) [inline src/models/sir.jl:2]: Drop this

File: src/models/sir.jl [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        migration_rates,
+        β_und,
+        β_det,
+        infection_period,
+        reinfection_probability,
+        detection_time,
+        death_rate,
+        Is,
+        seed
+    )
```
Comment (Libbum) [inline src/models/sir.jl:65]:
```julia
params = Dict(
        Ns => Ns,
       migration_rates  => migration_rates,
       β_und =>  β_und,
       β_det  => β_det,
        infection_period  => infection_period,
       reinfection_probability  => reinfection_probability,
        detection_time => detection_time,
       death_rate  => death_rate,
       ls => Is,
       seed => seed
    )```

**Comment by Libbum** [comment]
@Datseris, but these are in the models module, not the docs.

**Comment by kavir1698** [comment]
> One question about https://github.com/fbanning/Agents.jl/blob/Extend-Models-submodule/src/models/wright-fisher.jl :
Is it really necessary to have those two functions with different names *_neutral and *_selection ? I think it might be possible to just define the function twice with the same name - once with and once without :traitas a parameter.
If that's correct, this could simplify the module a bit and might demonstrate usage of methods/multiple dispatch for stepping functions.

"model step" functions have always had a single argument, the model object. I think it is better not to break this convention in this example.

**Comment by Datseris** [comment]
> @Datseris, but these are in the models module, not the docs.

Sorry , you are right, and I agree with your suggestion to drop it there.

**Comment by fbanning** [comment]
Ok, I will apply those changes in a moment.

**Comment by Libbum** [comment]
@fbanning, is it possible for you to enable "Allow edits from maintainers" on your `Extend-Models-submodule` branch? It would help speed up these travis conflicts.

The next hurdle is
> invalid redefinition of constant PoorSoul

The sir and social distancing models both have the same struct name. I'd suggest keeping the name of the one in the sir model and changing the name in social distancing. Perhaps `SocietyMember`?

**Comment by fbanning** [comment]
grafik

It's already enabled?

**Comment by Libbum** [comment]
Cool, thanks. I'll try again if something else fails :)

**Comment by Libbum** [comment]
I'm going to do some offline work to try and get this building. Will push up some fixes soon.

**Comment by Libbum** [comment]
That should fix the build errors. Docs errors may still happen though...

**Comment by Libbum** [comment]
Things to finalise here.

- [ ] Wealth model is the 1D version, in the example a 2D version also exists. @Datseris, @kavir1698: Should we add the 2D version into the model library or just state in the example file the 1D version is available in the model library only?
- [x] SIR model needs lightgraphs. Hopefully my last commit fixes that.
- [x] HK needs StatsBase for mean. As above.
- [ ] Should we rename `HK()` to `hk()`? The SIR model is `sir()`...
- [x] Types for Daisyworld and Predator-prey need to be exported. Fixed
- [x] SimpleCell needs to be fixed in the model and the example. 

Perhaps more. I'll update this section if and when I find them. Want to make sure these changes are correct first.

**Comment by Datseris** [comment]
I would actually put the 2D version of the wealth model in the `Models`. :D

**Comment by Libbum** [comment]
OK, I've gone through and verified all the changes work in the documentation, and validated that using the `Model` generated `model, agent_step!, model_step!` values can re-generate results from the example texts.

Where that hasn't been the case is in my comment above.

@fbanning: The Wealth Model changes can be done now, although the other issues will need comments from the other maintainers before we proceed.

That being said, this is essentially done now - most of this looks like minor changes and design choice.

**Comment by Datseris** [comment]
* Test coverage is meaningless to me, since it counts whether a functions is **called** at the tests. Whether it is tested properly is a completely different story, so I just don't care about coverages.
* yes for using `reverse(sincos(agent.orientation))`
* I would be happy to have a game of life in `Models`.
* The examples explicitly define `Daisy` themselves. Maybe in the docs of e.g. `daisyworld` we should just say "use `using Agents.Models: Daisy, Land` to access agent types".
* Yes for naming all models with lower-case.

**Comment by Libbum** [comment]
Having issues pushing changes at the moment, but all of these design decisions are taken care of.
@fbanning: I'll leave the [game of life](https://github.com/JuliaDynamics/Agents.jl/blob/master/examples/game_of_life_2D_CA.jl) inclusion up to you? Exactly the same as the other ones!

**Comment by Libbum** [comment]
@kavir1698, @Datseris: Either of you have any more comments on this?

**Comment by kavir1698** [comment]
Looks good to me

**Comment by Datseris** [comment]
to me too

**Comment by Libbum** [comment]
Nice catch on that last bit @fbanning. Thanks for your help on this one - not to complex, but a lot of finicky changes. Very much appreciated!

**Comment by fbanning** [comment]
Glad I could help! :)

**Pull request merged**